File: examples/practical-app/tests/e2e/modal.spec.ts
import { test, expect } from './fixtures';

test.describe('Modal Functionality', () => {
  test.beforeEach(async ({ page }) => {
    await page.goto('http://localhost:3001/');
    await page.waitForLoadState('networkidle');

    // Clear localStorage to start fresh
    await page.evaluate(() => localStorage.clear());
    await page.reload();
    await page.waitForLoadState('networkidle');

    // Add a todo so we can edit it to trigger the modal
    await page.locator('[data-todo-target="input"]').fill('Test todo for modal');
    await page.keyboard.press('Enter');
    await expect(page.locator('[data-todo-target="list"]').locator('text=Test todo for modal')).toBeVisible();
  });

  // Helper to open modal via edit button
  async function openModal(page: import('@playwright/test').Page) {
    const editButton = page.locator('[data-action*="todo#edit_todo"]').first();
    await editButton.click();
    // Wait for modal to become active
    await expect(page.locator('.modal.active')).toBeVisible();
  }

  test('should open modal', async ({ page }) => {
    const modal = page.locator('.modal[data-controller="modal"]');

    // Modal should be hidden initially (no active class)
    await expect(modal).not.toHaveClass(/active/);

    // Click edit button to open modal
    await openModal(page);

    // Modal should be visible (has active class)
    await expect(modal).toHaveClass(/active/);
  });

  test('should close modal with close button', async ({ page }) => {
    const modal = page.locator('.modal[data-controller="modal"]');

    // Open modal
    await openModal(page);
    await expect(modal).toHaveClass(/active/);

    // Click the X close button (use specific selector to avoid matching overlay)
    const closeButton = page.locator('button.modal-close');
    await closeButton.click();

    // Modal should be hidden
    await expect(modal).not.toHaveClass(/active/);
  });

  test('should close modal with Cancel button', async ({ page }) => {
    const modal = page.locator('.modal[data-controller="modal"]');

    // Open modal
    await openModal(page);
    await expect(modal).toHaveClass(/active/);

    // Click Cancel button (secondary close option)
    await page.locator('.modal button:has-text("Cancel")').click();

    // Modal should be hidden
    await expect(modal).not.toHaveClass(/active/);
  });

  test('should close modal with Escape key', async ({ page }) => {
    const modal = page.locator('.modal[data-controller="modal"]');

    // Open modal
    await openModal(page);
    await expect(modal).toHaveClass(/active/);

    // Press Escape
    await page.keyboard.press('Escape');

    // Modal should be hidden
    await expect(modal).not.toHaveClass(/active/);
  });

  test('should prevent body scroll when modal is open', async ({ page }) => {
    // Open modal
    await openModal(page);

    // Check if body has overflow hidden
    const bodyOverflow = await page.evaluate(() => {
      return window.getComputedStyle(document.body).overflow;
    });

    expect(bodyOverflow).toBe('hidden');
  });

  test('should restore body scroll when modal closes', async ({ page }) => {
    const modal = page.locator('.modal[data-controller="modal"]');

    // Open modal
    await openModal(page);

    // Close modal using the X button
    const closeButton = page.locator('button.modal-close');
    await closeButton.click();

    // Wait for modal to close
    await expect(modal).not.toHaveClass(/active/);

    // Body overflow should be restored
    const bodyOverflow = await page.evaluate(() => {
      return window.getComputedStyle(document.body).overflow;
    });

    expect(['visible', 'auto', '']).toContain(bodyOverflow);
  });

  test('should focus input within modal', async ({ page }) => {
    // Open modal
    await openModal(page);

    // Wait for focus to be set (modal uses setTimeout for focus)
    await page.waitForTimeout(200);

    // Focus should be on the input inside modal
    const focusedElement = await page.evaluate(() => {
      return document.activeElement?.closest('.modal[data-controller="modal"]') !== null;
    });

    expect(focusedElement).toBe(true);
  });

  test('should save changes when save button is clicked', async ({ page }) => {
    const modal = page.locator('.modal[data-controller="modal"]');

    // Open modal
    await openModal(page);

    // Change the todo text
    const modalInput = page.locator('[data-modal-target="input"]');
    await modalInput.clear();
    await modalInput.fill('Updated todo text');

    // Click save button
    await page.locator('.modal button:has-text("Save")').click();

    // Modal should close
    await expect(modal).not.toHaveClass(/active/);

    // Todo should be updated
    await expect(page.locator('[data-todo-target="list"]').locator('text=Updated todo text')).toBeVisible();
  });
});

File: examples/practical-app/spec/features/modal_spec.rb
# frozen_string_literal: true

RSpec.describe 'Modal Functionality', type: :feature do
  let(:modal_selector) { '.modal[data-controller="modal"]' }
  let(:input_selector) { '[data-todo-target="input"]' }
  let(:list_selector) { '[data-todo-target="list"]' }

  before do
    # Add a todo so we can edit it to trigger the modal
    input = find(input_selector, wait: 10)
    input.set('Test todo for modal')
    # Give browser time to process the input value
    sleep 0.3
    input.native.send_keys(:enter)
    # Wait for Opal to process the event
    sleep 1.0
    # Wait for todo to be added
    expect(page).to have_css("#{list_selector} .todo-item", wait: 10)
    expect(page).to have_content('Test todo for modal')
  end

  def open_modal
    find('[data-action*="todo#edit_todo"]', match: :first).click
    expect(page).to have_css('.modal.active', wait: 5)
  end

  def modal_element
    # Find modal even if hidden
    find(modal_selector, visible: :all)
  end

  describe 'opening modal' do
    it 'opens modal' do
      modal = modal_element
      expect(modal[:class]).not_to include('active')

      open_modal

      # Re-find the modal after it becomes active
      modal = find(modal_selector)
      expect(modal[:class]).to include('active')
    end
  end

  describe 'closing modal' do
    before { open_modal }

    it 'closes modal with close button' do
      modal = find(modal_selector)
      expect(modal[:class]).to include('active')

      find('button.modal-close').click

      expect(modal[:class]).not_to include('active')
    end

    it 'closes modal with Cancel button' do
      modal = find(modal_selector)

      within(modal_selector) do
        click_button 'Cancel'
      end

      expect(modal[:class]).not_to include('active')
    end

    it 'closes modal with Escape key' do
      modal = find(modal_selector)

      find('body').send_keys(:escape)

      expect(modal[:class]).not_to include('active')
    end
  end

  describe 'body scroll behavior' do
    it 'prevents body scroll when modal is open' do
      open_modal

      body_overflow = page.evaluate_script("window.getComputedStyle(document.body).overflow")
      expect(body_overflow).to eq('hidden')
    end

    it 'restores body scroll when modal closes' do
      open_modal

      find('button.modal-close').click

      # Wait for modal to close
      expect(page).not_to have_css('.modal.active')

      body_overflow = page.evaluate_script("window.getComputedStyle(document.body).overflow")
      expect(['visible', 'auto', '']).to include(body_overflow)
    end
  end

  describe 'focus management' do
    it 'focuses input within modal' do
      open_modal

      # Wait for focus to be set
      sleep 0.2

      focused_in_modal = page.evaluate_script(
        "document.activeElement?.closest('.modal[data-controller=\"modal\"]') !== null"
      )
      expect(focused_in_modal).to be true
    end
  end

  describe 'saving changes' do
    it 'saves changes when save button is clicked' do
      open_modal

      # Change the todo text
      modal_input = find('[data-modal-target="input"]')
      modal_input.set('Updated todo text')

      within(modal_selector) do
        click_button 'Save'
      end

      # Modal should close
      expect(page).not_to have_css('.modal.active')

      # Todo should be updated
      within(list_selector) do
        expect(page).to have_content('Updated todo text')
      end
    end
  end
end
